Changes to support forward references.

bader opened the pull request:
This change is required for enabling forward reference support used list data structure.

struct Node {
    Node *next;
    int value;
};

bader committed:
Changes to support forward references.

This change is required for enabling forward reference support used list data structure.

struct Node {
    Node *next;
	int value;
};

yxsamliu commented on lib/SPIRV/libSPIRV/SPIRVType.cpp:
```diff
@@ -240,6 +240,27 @@ SPIRVType::isTypeVectorOrScalarFloat() const {
+  Decoder >> Id;
+
+  for (size_t i = 0, e = MemberTypeVec.size(); i != e; ++i) {
+    SPIRVId currId;
```
currId should be capitalized for LLVM style, i and e may be renamed as MI and ME

yxsamliu commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -380,6 +380,37 @@ LLVMToSPIRV::isBuiltinTransToExtInst(Function *F,
+static bool recursiveType(const StructType *ST, const Type *Ty) {
+  SmallPtrSet<const StructType *, 4> Seen;
+
+  std::function<bool(const Type *Ty)> run = [&](const Type *Ty) {
```
run should be capitalized for LLVM style

yxsamliu commented on test/SPIRV/transcoding/RecursiveType.ll:
```diff
@@ -0,0 +1,43 @@
+target datalayout = "e-p:32:32-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024"
+target triple = "spir-unknown-unknown"
+
+%struct.Node = type { %struct.Node addrspace(1)*, i32 }
```
Better to add a test for nested structs, e.g.

struct A;
struct B {
  int b1;
  struct A *b2;
};
struct C {
  int c1;
  struct B c2;
};
struct A {
  int a1;
  struct C a2;
};
global struct A a;

the IR is:
%struct.A = type { i32, %struct.C }
%struct.C = type { i32, %struct.B }
%struct.B = type { i32, %struct.A\* }

@a = common global %struct.A zeroinitializer, align 8

bader replied:
Added similar case to the test.
Also I've added support for OpTypeForwardPointer as it's required by SPIR-V specification: https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#_universal_validation_rules (data rules).

bader committed:
Merge remote-tracking branch 'github/khronos/spirv-3.6.1' into khronos/list-data-structure-fix

bader committed:
[SPIR-V] Added recursive type definition support.

Applied code review from suggestions from Sam.
Added support for TypeForwardPointer required by SPIR-V specification.

yxsamliu commented:
LGTM. Please resolve conflicts.

bader committed:
Merge remote-tracking branch 'github/khronos/spirv-3.6.1' into khronos/list-data-structure-fix

Conflicts:
	lib/SPIRV/SPIRVWriter.cpp

bader commented:
Merged with the latest version of khronos/spirv-3.6.1 branch.
